### PR TITLE
Add admin resource list

### PR DESCRIPTION
**Pull request by asalkeld**
tested with https://msazure.visualstudio.com/One/_git/Compute-ARO-GenevaActions/pullrequest/2764297?_a=overview
I have separated out the oc apply and will post that next.

### CODE REVIEW

File: pkg/frontend/admin_openshiftcluster_resources_list.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+	doc, err := f.db.OpenShiftClusters.Get(ctx, resourceID)
+	switch {
+	case cosmosdb.IsErrorStatusCode(err, http.StatusNotFound):
+		return nil, api.NewCloudError(http.StatusNotFound, api.CloudErrorCodeResourceNotFound, "", "The Resource '%s/%s' under resource group '%s' was not found.", vars["resourceType"], vars["resourceName"], vars["resourceGroupName"])
+	case err != nil:
+		return nil, err
+	}
+
+	fpAuthorizer, err := f.env.FPAuthorizer(f.env.TenantID(), azure.PublicCloud.ResourceManagerEndpoint)
```
Comment (asalkeld) [inline pkg/frontend/admin_openshiftcluster_resources_list.go:46]: @jim-minter I will move this, I am just not sure on what authorizer to use.
This is currently not returning what i expect, it returns the resources, but with very limited info, but if do
```az resource list -g v4-angus-rg -o table --debug```
then each resource is properly populated, I am guessing that it is the difference in auth? - as the sent url is the same in both cases.

File: pkg/frontend/kubeactions/kubeactions.go [hunk 2]
```diff
@@ -160,6 +163,61 @@ func (ka *kubeactions) List(ctx context.Context, oc *api.OpenShiftCluster, kind,
 	return ul.MarshalJSON()
 }
 
+func (ka *kubeactions) applyOne(ctx context.Context, dyn dynamic.Interface, grs []*restmapper.APIGroupResources, un *unstructured.Unstructured) error {
+	if strings.EqualFold(un.GetKind(), "secret") {
```
Comment (asalkeld) [inline pkg/frontend/kubeactions/kubeactions.go:167]: is this needed? or is it just not viewing secrets..

File: pkg/frontend/admin_openshiftcluster_resources_list.go (hunk 1 above)
Comment (jim-minter) [reply pkg/frontend/admin_openshiftcluster_resources_list.go:46]: I believe this is correct.

Comment (jim-minter) [reply pkg/frontend/admin_openshiftcluster_resources_list.go:46]: I don't think there's an auth difference.  I think you would need to iterate the resources returned by list, then get each individually.  You should be able to use some resources client to do that (maybe even the one you've got).  It's a bit like the Azure equivalent of the k8s dynamic API.

Comment (asalkeld) [reply pkg/frontend/admin_openshiftcluster_resources_list.go]: Thanks @jim-minter this works nicely now.

File: pkg/frontend/admin_openshiftcluster_resources_list_test.go [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+		resources:       mr,
+		bucketAllocator: &bucket.Random{},
+	}
+	f.ocEnricher = clusterdata.NewBestEffortEnricher(f.baseLog, _env)
+	err := f.setupListener(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return f, nil
+}
```
Comment (jim-minter) [inline pkg/frontend/admin_openshiftcluster_resources_list_test.go:54]: please can you put this in an independent file, perhaps in frontend_test.go ?

**Comment by asalkeld** [comment]
> Please also check the CI failure.

Thanks for the review @jim-minter . I hope I have gotten all the changes..
Note:
- I have merged the apiVersions (nothing conflicted) and I now generate them from the godep.lock
- I now use this merged apiVersions to figure out the correct version to pass into GetByID(). It does really give heaps of info now, so it is working better.

File: pkg/util/azureclient/mgmt/features/resources_addons.go [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+import (
+	"context"
+
+	mgmtfeatures "github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2019-07-01/features"
+)
+
+// ResourcesClientAddons is a minimal interface for azure ResourcesClient
+type ResourcesClientAddons interface {
+	List(ctx context.Context, filter, expand string, top *int32) (mgmtfeatures.ResourceListResult, error)
```
Comment (jim-minter) [inline pkg/util/azureclient/mgmt/features/resources_addons.go:14]: please return `[]mgmtfeatures.GenericResourceExpanded, error` for consistency

File: pkg/util/azureclient/typeToCode.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
+package azureclient
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import "strings"
+
+// GetCodeFromType gets the code (that keys APIVersions) from a resource type
+func GetCodeFromType(_type string) string {
```
Comment (jim-minter) [inline pkg/util/azureclient/typeToCode.go:9]: I agree that we need to be able to map from a received _type to a version, but I'd rather not do it like this.  Can you refactor so that APIVersions looks like something this, then work from there?

```
var APIVersions = map[string]string{
	"Microsoft.Authorization":                 "2018-09-01-preview",
	"Microsoft.Authorization/denyAssignments": "2018-07-01-preview",
	"Microsoft.Authorization/roleDefinitions": "2018-01-01-preview",
	"Microsoft.Compute":                       "2019-03-01",
	"Microsoft.ContainerRegistry":             "2019-05-01",
	"Microsoft.DocumentDB":                    "2019-08-01",
	"Microsoft.Insights":                      "2018-03-01",
	"Microsoft.KeyVault":                      "2016-10-01",
	"Microsoft.ManagedIdentity":               "2018-11-30",
	"Microsoft.Network":                       "2019-07-01",
	"Microsoft.Network/dnsZones":              "2018-05-01",
	"Microsoft.Network/privateDnsZones":       "2018-09-01",
	"Microsoft.Storage":                       "2019-04-01",
}
```

File: pkg/frontend/frontend.go [hunk 6]
```diff
@@ -59,7 +62,11 @@ type Runnable interface {
 
 // NewFrontend returns a new runnable frontend
 func NewFrontend(ctx context.Context, baseLog *logrus.Entry, _env env.Interface, db *database.Database, apis map[string]*api.Version, m metrics.Interface, kubeActions kubeactions.Interface) (Runnable, error) {
-	var err error
+	fpAuthorizer, err := _env.FPAuthorizer(_env.TenantID(), azure.PublicCloud.ResourceManagerEndpoint)
+	if err != nil {
+		return nil, err
+	}
+	resourcesClient := features.NewResourcesClient(_env.SubscriptionID(), fpAuthorizer)
```
Comment (jim-minter) [inline pkg/frontend/frontend.go:69]: as discussed on Slack, this needs to be built after the request has come in.

File: pkg/util/azureclient/mgmt/features/resources_addons.go (hunk 4 above)
Comment (jim-minter) [reply pkg/util/azureclient/mgmt/features/resources_addons.go:14]: bump

File: pkg/util/azureclient/typeToCode.go [hunk 7]
```diff
@@ -0,0 +1,32 @@
+package azureclient
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import "strings"
+
+// GetAPIVersionFromType gets the APIVersion from a full resource type
+func GetAPIVersionFromType(_type string) string {
+	defaultVer := APIVersions["Microsoft.Compute"]
```
Comment (jim-minter) [inline pkg/util/azureclient/typeToCode.go:10]: why is this needed and how can it help?

File: pkg/util/azureclient/typeToCode.go [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+	if strings.Contains(_type, "/providers/") {
+		// this is to account for types like this: Microsoft.Network/dnsZones/providers/roleAssignments
+		// that needs to be Microsoft.Authorization not Microsoft.Network
```
Comment (jim-minter) [inline pkg/util/azureclient/typeToCode.go:13]: what are the circumstances where this is needed?

File: pkg/util/azureclient/typeToCode.go (hunk 7 above)
Comment (asalkeld) [reply pkg/util/azureclient/typeToCode.go:10]: well what are the other options? return an empty string/error 
I was hoping that at least it would glean some info - better than nothing

File: pkg/util/azureclient/typeToCode.go (hunk 8 above)
Comment (asalkeld) [reply pkg/util/azureclient/typeToCode.go:13]: i'll remove as see what happens

File: pkg/util/azureclient/typeToCode.go (hunk 7 above)
Comment (jim-minter) [reply pkg/util/azureclient/typeToCode.go]: if we have an unknown resource type, there is no guarantee that the compute api version will work for it.  We should find a different mechanism to indicate this fault.  A log line + leaving the item unenriched should be OK, shouldn't it?

**Comment by jim-minter** [comment]
pushed to restructure commits and rebase.  In retrospect this was too much for one PR.
